README.md
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/onyx-and-iris/voicemeeter-api-ruby/blob/dev/LICENSE)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/plugin-ruby)
![tests: failed count](https://img.shields.io/badge/dynamic/json?color=blue&label=tests&query=summary.failure_count&suffix=%20failed&url=https%3A%2F%2Fraw.githubusercontent.com%2Fonyx-and-iris%2Fvoicemeeter-api-ruby%2Fdev%2Fspec%2Frspec.json)
[![Gem Version](https://badge.fury.io/rb/voicemeeter_api_ruby.svg)](https://badge.fury.io/rb/voicemeeter_api_ruby)
![](https://ruby-gem-downloads-badge.herokuapp.com/voicemeeter_api_ruby?type=total&color=red)

# Ruby Wrapper for Voicemeeter API

This gem offers a Ruby interface for the Voicemeeter Remote C API.

For an outline of past/future changes refer to: [CHANGELOG](CHANGELOG.md)

## Tested against

-   Basic 1.0.8.2
-   Banana 2.0.6.2
-   Potato 3.0.2.2

## Requirements

-   [Voicemeeter](https://voicemeeter.com/)
-   Ruby 3.0 or greater

## Installation

### Bundler

Put this in your Gemfile:

`gem 'voicemeeter_api_ruby'`

### Gem

Install voicemeeter_api_ruby gem from your console

`gem install 'voicemeeter_api_ruby'`

## `Use`

Simplest use case, request a Remote class of a kind, then pass a block to run.

Login and logout are handled for you in this scenario.

#### `main.rb`

```ruby
require 'voicemeeter'

kind_id = 'banana'

# start Voicemeeter GUI
Voicemeeter.start(kind_id)

vm = Voicemeeter.remote(kind_id)

# vm.run accepts a block
vm.run do
    # mute the leftmost strip
    vm.strip[0].mute = true
    puts vm.strip[0].mute

    # disable eq for second from left bus
    vm.bus[1].eq = false
    puts vm.bus[1].eq
end
```

Otherwise you must remember to call `vm.login` `vm.logout` at the start/end of your code.

## `kind_id`

Pass the kind of Voicemeeter as an argument. kind_id may be:

-   `basic`
-   `banana`
-   `potato`

## `Available commands`

### Strip

The following properties are available.

-   `mono`: boolean
-   `solo`: boolean
-   `mute`: boolean
-   `gain`: float, from -60.0 to 12.0
-   `comp`: float, from 0.0 to 10.0
-   `gate`: float, from 0.0 to 10.0
-   `audibility`: float, from 0.0 to 10.0
-   `limit`: int, from -40 to 12
-   `A1 - A5`, `B1 - B3`: boolean
-   `label`: string
-   `device`: string
-   `sr`: int
-   `mc`: boolean
-   `k`: int, from 0 to 4
-   `bass`: float, from -12.0 to 12.0
-   `mid`: float, from -12.0 to 12.0
-   `treble`: float, from -12.0 to 12.0
-   `reverb`: float, from 0.0 to 10.0
-   `delay`: float, from 0.0 to 10.0
-   `fx1`: float, from 0.0 to 10.0
-   `fx2`: float, from 0.0 to 10.0
-   `pan_x`: float, from -0.5 to 0.5
-   `pan_y`: float, from 0.0 to 1.0
-   `color_x`: float, from -0.5 to 0.5
-   `color_y`: float, from 0.0 to 1.0
-   `fx_x`: float, from -0.5 to 0.5
-   `fx_y`: float, from 0.0 to 1.0
-   `postreverb`: boolean
-   `postdelay`: boolean
-   `postfx1`: boolean
-   `postfx2`: boolean

example:

```ruby
vm.strip[3].gain = 3.7
puts vm.strip[0].label
```

The following methods are available.

-   `appgain(name, value)`: string, float, from 0.0 to 1.0

Set the gain in db by value for the app matching name.

-   `appmute(name, value)`: string, bool

Set mute state as value for the app matching name.

example:

```ruby
vm.strip[5].appgain('Spotify', 0.5)
vm.strip[5].appmute('Spotify', true)
```

##### Gainlayers

-   `gain`: float, from -60.0 to 12.0

example:

```ruby
vm.strip[3].gainlayer[3].gain = 3.7
```

Gainlayers are defined for potato version only.

##### Levels

The following properties are available.

-   `prefader`
-   `postfader`
-   `postmute`

example:

```ruby
puts vm.strip[3].levels.prefader
```

Level properties will return -200.0 if no audio detected.

### Bus

The following properties are available.

-   `mono`: boolean
-   `mute`: boolean
-   `eq`: boolean
-   `eq_ab`: boolean
-   `sel`: boolean
-   `gain`: float, from -60.0 to 12.0
-   `label`: string
-   `device`: string
-   `sr`: int
-   `returnreverb`: float, from 0.0 to 10.0
-   `returndelay`: float, from 0.0 to 10.0
-   `returnfx1`: float, from 0.0 to 10.0
-   `returnfx2`: float, from 0.0 to 10.0
-   `monitor`: boolean

example:

```ruby
vm.bus[3].gain = 3.7
puts vm.bus[0].label

vm.bus[4].mono = true
```

##### Modes

-   `normal`: boolean
-   `amix`: boolean
-   `bmix`: boolean
-   `composite`: boolean
-   `tvmix`: boolean
-   `upmix21`: boolean
-   `upmix41`: boolean
-   `upmix61`: boolean
-   `centeronly`: boolean
-   `lfeonly`: boolean
-   `rearonly`: boolean

example:

```ruby
vm.bus[4].mode.amix = true
```

##### Levels

The following properties are available.

-   `all`

example:

```ruby
puts vm.bus[0].levels.all
```

`levels.all` will return -200.0 if no audio detected.

### Strip | Bus

The following methods are Available

-   `fadeto(amount, time)`: float, int
-   `fadeby(amount, time)`: float, int

Modify gain to or by the selected amount in db over a time interval in ms.

example:

```ruby
vm.strip[0].fadeto(-10.3, 1000)
vm.bus[3].fadeby(-5.6, 500)
```

### Macrobuttons

Three modes defined: state, stateonly and trigger.

-   `state`: boolean
-   `stateonly`: boolean
-   `trigger`: boolean

example:

```ruby
vm.button[37].state = true
vm.button[55].trigger = false
```

### Recorder

The following properties accept boolean values.

-   `loop`: boolean
-   `A1 - A5`: boolean
-   `B1 - A3`: boolean

The following methods are Available

-   `play`
-   `stop`
-   `pause`
-   `record`
-   `ff`
-   `rew`
-   `load(filepath)`: string

example:

```ruby
vm.recorder.play
vm.recorder.stop

# Enable loop play
vm.recorder.loop = true

# Disable recorder out channel B2
vm.recorder.B2 = false

# filepath as string
vm.recorder.load('C:\music\mytune.mp3')
```

### VBAN

-   `vm.vban.enable` `vm.vban.disable` Turn VBAN on or off

##### Instream | Outstream

The following properties are available.

-   `on`: boolean
-   `name`: string
-   `ip`: string
-   `port`: int, range from 1024 to 65535
-   `sr`: int, (11025, 16000, 22050, 24000, 32000, 44100, 48000, 64000, 88200, 96000)
-   `channel`: int, from 1 to 8
-   `bit`: int, 16 or 24
-   `quality`: int, from 0 to 4
-   `route`: int, from 0 to 8

SR, channel and bit are defined as readonly for instreams. Attempting to write to those parameters will throw an error. They are read and write for outstreams.

example:

```ruby
# turn VBAN on
vm.vban.enable

# turn on vban instream 0
vm.vban.instream[0].on = true

# set bit property for outstream 3 to 24
vm.vban.outstream[3].bit = 24
```

### Command

Certain 'special' commands are defined by the API as performing actions rather than setting values. The following methods are available:

-   `show` : Bring Voiceemeter GUI to the front
-   `shutdown` : Shuts down the GUI
-   `restart` : Restart the audio engine

The following properties are write only and accept boolean values.

-   `showvbanchat`: boolean
-   `lock`: boolean

example:

```ruby
vm.command.restart
vm.command.showvbanchat = true
```

### Device

-   `ins` `outs` : Returns the number of input/output devices
-   `input(i)` `output(i)` : Returns a hash of device properties for device[i]

example:

```ruby
vm.run { (0...vm.device.ins).each { |i| puts vm.device.input(i) } }
```

### Multiple parameters

-   `apply`
    Set many strip/bus/macrobutton/vban parameters at once, for example:

```ruby
vm.apply(
    {
        strip_0: {
            mute: true,
            gain: 3.2,
            A1: true,
        },
        bus_3: {
            gain: -3.2,
            eq: true,
        },
        button_39: {
            stateonly: true,
        },
        vban_outstream_3: {
            on: true,
            bit: 24,
        },
    },
)
```

Or for each class you may do:

```ruby
vm.strip[0].apply(mute: true, gain: 3.2, A1: true)
vm.vban.outstream[0].apply(on: true, name: 'streamname', bit: 24)
```

## Config Files

`vm.apply_config(<configname>)`

You may load config files in TOML format.
Three example configs have been included with the package. Remember to save
current settings before loading a config. To set one you may do:

```ruby
require 'voicemeeter'
vm = Voicemeeter.remote('banana')
vm.run { vm.apply_config('example') }
```

will load a config file at mydir/configs/banana/example.toml for Voicemeeter Banana.

## `Voicemeeter Module`

### Remote class

Access to lower level Getters and Setters are provided with these functions:

-   `vm.get(param, string=false)`: For getting the value of any parameter. Set string to true if getting a property value expected to return a string.
-   `vm.set(param, value)`: For setting the value of any parameter.

Access to lower level polling functions are provided with these functions:

-   `vm.pdirty?`: Returns true if a parameter has been updated.
-   `vm.mdirty?`: Returns true if a macrobutton has been updated.

example:

```ruby
vm.get('Strip[2].Mute')
vm.set('Strip[4].Label', 'stripname')
vm.set('Strip[0].Gain', -3.6)
```

#### Voicemeeter::start

Use this function to start Voicemeeter of a kind independently of Remote class.
for example:

```ruby
require 'voicemeeter'
Voicemeeter.start('banana')
```

### Run tests

To run all tests:

```
Bundle exec rspec
```

You can use tags to run only certain tests, for example:

```
Bundle exec rspec --tag 'higher'
```

### Official Documentation

-   [Voicemeeter Remote C API](https://github.com/onyx-and-iris/Voicemeeter-SDK/blob/main/VoicemeeterRemoteAPI.pdf)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Before any major/minor/patch is released all unit tests will be run to verify they pass.

## [Unreleased] - These changes have not been added to RubyGems yet

-   [ ]

## [4.3.0] - 2022-07-29

### Added

-   Mixin module
-   xy, fx parameters to strip/bus
-   unit tests for xy, fx
-   xy, fx to readme
-   config tests added
-   gemfile to obs example

### Changed

-   moved gem files into voicemeeter/ dir
-   update obs example
-   using require_relative in case using direct download.

## [4.1.0] - 2022-07-19

### Added

-   Conversion module in meta
-   strip/bus levels implemented, ldirty added to notifications.

### Changed

-   1, 0 switched for true false in example configs
-   observer example updated.
-   minor version bump

### Removed

-   Boolean class monkey patch
-   type checks

## [4.0.0] - 2022-06-27

### Added

-   type, version added to base class
-   device class implemented
-   TOMLStrBuilder added to config, builds a default reset config
-   support for observers to subscribe to updates
-   observer example added
-   device, sr, sel added to bus
-   audibility, bass, mid, treble added to strip

### Changed

-   common interface iremote defined
-   configs reworked. configs are now loaded lazily.
-   command.reset now applies reset config from toml parser.
-   profiles dir renamed to configs
-   major version bump due to non-backwards compatible changes

## [3.0.0] - 2022-05-03

### Added

-   Added strip_levels, bus_levels to base class. They return the entire level array (before math conversion).
-   Added fadeto, fadeby methods for strips/buses.
-   Added appgain, appmute methods to Virtualstrip class
-   Bus modes added to bus class.
-   Gainlayers added to strip class.
-   Added pre-commit.ps1 for git hook
-   cdll proc object added to base class. This is now used to call a c dll function.
-   hide method added to command class

### Changed

-   Kinds, Profiles moved into their own modules.
-   Kind maps now defined as structs.
-   Randomised unit tests. A single run may test any kind.
-   Polling paramters now defined as single line endless functions
-   Major version bump due to backwards incompatible changes. Ruby required version changed.

### Fixed

-   Minor delay added before logout, allow a final operation to complete.
-   Polling function sped up, sync kwarg added.

### Removed

-   Removed bounds checking for certain strip/bus properties.

## [2.0.4] - 2022-01-24

### Added

-   Ability to load config files in TOML format
-   Add option to pass keyword arguments to Remote class.
-   Voicemeeter.start added for starting voicemeeter independently of Remote class

### Changed

-   Use of VBVMR_RunVoicemeeter instead of open3 when starting voicemeeter.
-   Functions in low level modules set to private.
-   version 2.0.4 added to RubyGems.

### Fixed

-   Fix timing issue affecting macrobutton multi_set

## [2.0.3] - 2022-01-20

### Added

-   runmany.ps1 added. easy access to many test runs with logging and summary file.
-   link to official documentation in readme.
-   section regarding lower level functions, setters, getters and polling functions added to readme.

### Changed

-   Adjustment to polling function, now accepts getter logic as a block. Param caching
    now done in polling function
-   Small delay added to setters, except for set_parameter_multi function.
-   Added initialize to Remote class, building of class objects done now done in Remote.
-   version 2.0.3 added to RubyGems.

### Fixed

-   Fixed an issue with the action type property meta function in Command class.
-   Added ensure logout if block passed to run, will logout regardless of exception raised.

## [2.0.2] - 2022-01-18

### IMPORTANT

The code for this wrapper has been revised and reworked from the ground up. Earlier versions
are still available for download on RubyGems.org but that version will not be updated further.
Most of the updates documented here do not apply to version 2.0 onwards but will be kept for
reference use only.

All of the core features including strip, bus, macrobutton, recorder, vban and command classes
work exactly the same. An effort has been made to include every property previously available,
for each of those classes.

README is up to date with version 2.0.2 onwards.

### Added

-   Special command lock added
-   showvbanchat chat is now a boolean function.
-   Functions vmr.vban.enable and vmr.vban.disable for toggling VBAN now added.

### Changed

-   Each class for each of the core features has been put into its own module. Classes such as
    strip, bus and vban are now subclassed into physical/virtual and instream/outstream respectively.
-   Error classes have been simplified into a smaller number of classes.
-   remote class is now callable through Voicemeeter module. It is also now a factory function that
    returns a Remote class for the kind of Voicemeeter specified by the user.
-   Rake tests now simplified and reduced. They are now organised into higher and lower tests.

### Fixed

-   Writer validation logic has been moved into meta module. There is a base module for generic
    meta functions but where appropriate this is mixed in with a submodule.
-   set_multi function for setting many parameters at once now uses tokens for hash keys.
-   Polling dirty parameters works similarly but the cache is purged upon successful fetch.

---

## [1.0.10] - 2021-05-12

### Added

-   More settable commands for VBAN parameters

### Changed

-   Update unit tests to reflect changes
-   Updated README to reflect changes.
-   define_types moved into utils.rb since it defines types for channels,
    macrobuttons and vban
-   number set int and real_number added to utils.rb to better define different

lib/base.rb
require "observer"

require_relative "runvm"
require_relative "configs"
require_relative "errors"

class Base
  "
  Base class responsible for wrapping the C Remote API

  Mixin required modules
  "
  include Observable
  include Configs
  include RunVM

  attr_accessor :strip, :bus, :button, :vban, :command, :recorder, :device
  attr_accessor :strip_mode

  attr_reader :kind, :p_in, :v_in, :p_out, :v_out, :retval, :cache
  attr_reader :running, :_strip_comp, :_bus_comp, :delay

  DELAY = 0.001
  SYNC = false
  RATELIMIT = 0.033
  SIZE = 1

  def initialize(kind, **kwargs)
    @kind = kind
    @p_in, @v_in = kind.layout[:strip].values
    @p_out, @v_out = kind.layout[:bus].values
    @cache = Hash.new
    @sync = kwargs[:sync] || SYNC
    @ratelimit = kwargs[:ratelimit] || RATELIMIT
    @running = false
    @strip_mode = 0
    @delay = DELAY
  end

  def init_thread
    @running = true
    @cache["strip_level"], @cache["bus_level"] = _get_levels
    Thread.new do
      loop do
        Thread.stop if !@running
        if pdirty?
          changed
          notify_observers("pdirty")
        end
        if mdirty?
          changed
          notify_observers("mdirty")
        end
        if ldirty?
          changed
          @_strip_comp =
            @cache["strip_level"].map.with_index do |x, i|
              !(x == @strip_buf[i])
            end
          @_bus_comp =
            @cache["bus_level"].map.with_index { |x, i| !(x == @bus_buf[i]) }
          @cache["strip_level"] = @strip_buf
          @cache["bus_level"] = @bus_buf
          notify_observers("ldirty")
        end
        sleep(@ratelimit)
      end
    end
  end

  def end_thread
    @running = false
  end

  def login
    @@cdll.call(:login)
    clear_polling
  rescue CAPIErrors => error
    case
    when error.value == 1
      self.start(@kind.name)
      clear_polling
    when error.value < 0
      raise
    end
  end

  def logout
    clear_polling
    sleep(0.1)
    @@cdll.call(:logout)
  end

  def type
    c_type = FFI::MemoryPointer.new(:long, SIZE)
    @@cdll.call(:vmtype, c_type)
    types = { 1 => "basic", 2 => "banana", 3 => "potato" }
    types[c_type.read_long]
  end

  def version
    c_ver = FFI::MemoryPointer.new(:long, SIZE)
    @@cdll.call(:vmversion, c_ver)
    v1 = (c_ver.read_long & 0xFF000000) >> 24
    v2 = (c_ver.read_long & 0x00FF0000) >> 16
    v3 = (c_ver.read_long & 0x0000FF00) >> 8
    v4 = c_ver.read_long & 0x000000FF
    "#{v1}.#{v2}.#{v3}.#{v4}"
  end

  def get_parameter(name, is_string = false)
    self.polling("get_parameter", name: name) do
      if is_string
        c_get = FFI::MemoryPointer.new(:string, 512, true)
        @@cdll.call(:get_parameter_string, name, c_get)
        c_get.read_string
      else
        c_get = FFI::MemoryPointer.new(:float, SIZE)
        @@cdll.call(:get_parameter_float, name, c_get)
        c_get.read_float.round(1)
      end
    end
  end

  def set_parameter(name, value)
    if value.is_a? String
      @@cdll.call(:set_parameter_string, name, value)
    else
      @@cdll.call(:set_parameter_float, name, value.to_f)
    end
    @cache.store(name, value)
  end

  def get_buttonstatus(id, mode)
    self.polling("get_buttonstatus", id: id, mode: mode) do
      c_get = FFI::MemoryPointer.new(:float, SIZE)
      @@cdll.call(:get_buttonstatus, id, c_get, mode)
      c_get.read_float.to_i
    end
  end

  def set_buttonstatus(id, state, mode)
    @@cdll.call(:set_buttonstatus, id, state, mode)
    @cache.store("mb_#{id}_#{mode}", state)
  end

  def set_parameter_multi(param_hash)
    param_hash.each do |(key, val)|
      prop, m2, m3, *rem = key.to_s.split("_")
      if m2.to_i.to_s == m2
        m2 = m2.to_i
      elsif m3.to_i.to_s == m3
        m3 = m3.to_i
      end

      case prop
      when "strip"
        self.strip[m2].set_multi(val)
      when "bus"
        self.bus[m2].set_multi(val)
      when "button", "mb"
        self.button[m2].set_multi(val)
      when "vban"
        if %w[instream in].include? m2
          self.vban.instream[m3].set_multi(val)
        elsif %w[outstream out].include? m2
          self.vban.outstream[m3].set_multi(val)
        end
      end
      sleep(DELAY)
    end
  end

  def get_level(type, index)
    c_get = FFI::MemoryPointer.new(:float, SIZE)
    @@cdll.call(:get_level, type, index, c_get)
    c_get.read_float
  end

  def _get_levels
    s = (0...(2 * @p_in + 8 * @v_in)).map { |i| get_level(@strip_mode, i) }
    b = (0...(8 * (@p_out + @v_out))).map { |i| get_level(3, i) }
    [s, b]
  end

  def get_num_devices(direction)
    unless %w[in out].include? direction
      raise VMRemoteErrors.new("expected in or out")
    end
    if direction == "in"
      val = @@cdll.call(:get_num_indevices)
    else
      val = @@cdll.call(:get_num_outdevices)
    end
    val[0]
  end

  def get_device_description(index, direction)
    unless %w[in out].include? direction
      raise VMRemoteErrors.new("expected in or out")
    end
    c_type = FFI::MemoryPointer.new(:long, SIZE)
    c_name = FFI::MemoryPointer.new(:string, 256, true)
    c_hwid = FFI::MemoryPointer.new(:string, 256, true)
    if direction == "in"
      @@cdll.call(:get_desc_indevices, index, c_type, c_name, c_hwid)
    else
      @@cdll.call(:get_desc_outdevices, index, c_type, c_name, c_hwid)
    end
    [c_name.read_string, c_type.read_long, c_hwid.read_string]
  end

  alias_method "set_multi", :set_parameter_multi
  alias_method "apply", :set_parameter_multi
  alias_method "get", :get_parameter
  alias_method "set", :set_parameter
  alias_method "pdirty", :pdirty?
  alias_method "mdirty", :mdirty?
  alias_method "ldirty", :ldirty?
end
